### PR TITLE
Dev: log_patterns: update patterns for pacemaker version 2.0+

### PR DESCRIPTION
**Pull request by liangxin1300**
## Problem
`crm history peinputs` and `crm history peinputs v` are failed to give outputs
```
15sp4-1:~ # crm history peinputs
INFO: Retrieving information from cluster nodes, please wait...
15sp4-1:~ # crm history peinputs v
INFO: Fetching updated logs from cluster nodes. Please wait...
Date       Start    End       Filename      Client     User       Origin      Tags
====       =====    ===       ========      ======     ====       ======      ====
```
And the debug info are meaningless like `DEBUG: +Event events: error`
```
15sp4-1:~ # crm -d history peinputs
2022/11/25 15:57:10
DEBUG: found pacemaker version: 2.1.2+20211124.ada5c3b36-150400.4.3.1
DEBUG: setting source to live
DEBUG: setting report times: <2022-11-25 06:57:10> - <None>
DEBUG: Found description.txt, created 1669362779.811167
DEBUG: history: setting year to 2022 (ts: 1669362779.811167)
DEBUG: Transition metadata loaded from /var/cache/crm/history-root/live/__meta.json
INFO: Fetching updated logs from cluster nodes. Please wait...
DEBUG: Candidate logs: ['/var/cache/crm/history-root/live/15sp4-1/ha-log.txt', '/var/cache/crm/history-root/live/15sp4-2/ha-log.txt']
DEBUG: Updating 15sp4-1 : /var/cache/crm/history-root/live/15sp4-1/ha-log.txt : /var/log/messages : 1663586
DEBUG: Updating 15sp4-2 : /var/cache/crm/history-root/live/15sp4-2/ha-log.txt : /var/log/messages : 1636294
DEBUG: updating /var/log/messages from 15sp4-1 (pos 1663586)
DEBUG: updating /var/log/messages from 15sp4-2 (pos 1636294)
DEBUG: found pacemaker version: 2.1.2+20211124.ada5c3b36-150400.4.3.1-2.1.2+20211124.ada5c3b36 in cib: /var/cache/crm/history-root/live/15sp4-2/cib.xml
DEBUG: parsing /var/cache/crm/history-root/live/15sp4-1/ha-log.txt
DEBUG: +Event node: 15sp4-1
DEBUG: +Event node: 15sp4-1
DEBUG: +Event node: 15sp4-1
DEBUG: +Event node: 15sp4-1
DEBUG: +Event events: error
DEBUG: +Event events: error
DEBUG: +Event events: error
DEBUG: parsing /var/cache/crm/history-root/live/15sp4-2/ha-log.txt
DEBUG: +Event node: 15sp4-2
DEBUG: +Event node: 15sp4-2
DEBUG: +Event node: 15sp4-2
DEBUG: +Event node: 15sp4-2
DEBUG: +Event node: 15sp4-2
DEBUG: +Event node: 15sp4-2
DEBUG: +Event node: 15sp4-2
DEBUG: +Event node: 15sp4-2
DEBUG: +Event node: 15sp4-2
DEBUG: +Event node: 15sp4-2
DEBUG: +Event node: 15sp4-2
DEBUG: +Event node: 15sp4-2
DEBUG: Transition metadata saved to /var/cache/crm/history-root/live/__meta.json
```
## Changes
After update patterns for pacemaker version 2.0+
```
15sp4-1:~ # crm history peinputs
INFO: Fetching updated logs from cluster nodes. Please wait...
/var/cache/crm/history-root/live/15sp4-1/pengine/pe-input-0.bz2
/var/cache/crm/history-root/live/15sp4-1/pengine/pe-input-1.bz2
/var/cache/crm/history-root/live/15sp4-1/pengine/pe-input-3.bz2
/var/cache/crm/history-root/live/15sp4-1/pengine/pe-input-4.bz2
/var/cache/crm/history-root/live/15sp4-1/pengine/pe-input-5.bz2
/var/cache/crm/history-root/live/15sp4-1/pengine/pe-input-6.bz2
/var/cache/crm/history-root/live/15sp4-1/pengine/pe-input-7.bz2
/var/cache/crm/history-root/live/15sp4-1/pengine/pe-input-9.bz2
/var/cache/crm/history-root/live/15sp4-1/pengine/pe-input-10.bz2
/var/cache/crm/history-root/live/15sp4-1/pengine/pe-input-11.bz2
/var/cache/crm/history-root/live/15sp4-1/pengine/pe-input-12.bz2

15sp4-1:~ # crm history peinputs v
INFO: Fetching updated logs from cluster nodes. Please wait...
Date       Start    End       Filename      Client     User       Origin      Tags
====       =====    ===       ========      ======     ====       ======      ====
2022-11-25 21:55:08 21:55:08  pe-input-0    crmd       hacluster  15sp4-1   
2022-11-25 21:55:09 21:55:09  pe-input-1    cibadmin   root       15sp4-1   
2022-11-25 21:58:49 21:58:49  pe-input-3    cibadmin   root       15sp4-2   
2022-11-25 21:58:51 22:28:51  pe-input-4    cibadmin   root       15sp4-2   
2022-11-25 22:33:08 22:33:08  pe-input-5    cibadmin   root       15sp4-2   
2022-11-25 22:33:19 22:33:19  pe-input-6    crm_node   root       15sp4-1   
2022-11-25 22:33:26 22:48:26  pe-input-7    cibadmin   root       15sp4-1   
2022-11-25 23:01:02 23:01:03  pe-input-9    cibadmin   root       15sp4-1   d st
2022-11-25 23:01:03 23:01:03  pe-input-10   cibadmin   root       15sp4-1   
2022-11-25 23:04:34 23:04:34  pe-input-11   cibadmin   root       15sp4-1   st
2022-11-25 23:04:36 23:04:36  pe-input-12   cibadmin   root       15sp4-1   st

```
And the debug messages look more sensible (part of debug messages)
```
DEBUG: +Event node: 15sp4-1: 2022-11-25T13:54:43.259862+08:00 15sp4-1 systemd[1]: Starting Corosync Cluster Engine...
DEBUG: +Event node: 15sp4-1: 2022-11-25T13:54:43.342952+08:00 15sp4-1 corosync[4942]:   [MAIN  ] Corosync Cluster Engine ('2.4.5'): started and ready to provide service.
DEBUG: +Event node: 15sp4-1: 2022-11-25T13:54:44.051041+08:00 15sp4-1 corosync[4937]: Starting Corosync Cluster Engine (corosync): [  OK  ]
DEBUG: +Event node: 15sp4-1: 2022-11-25T13:54:44.051329+08:00 15sp4-1 systemd[1]: Started Corosync Cluster Engine.
DEBUG: +Event node: : 2022-11-25T13:54:46.883050+08:00 15sp4-1 pacemaker-controld[4957]:  notice: Fencer successfully connected
DEBUG: +Event quorum: quorum: 2022-11-25T13:55:07.024503+08:00 15sp4-1 pacemaker-controld[4957]:  notice: Updating quorum status to true (call=26)
DEBUG: +Event events: error: 2022-11-25T13:55:08.072042+08:00 15sp4-1 pacemaker-schedulerd[4956]:  error: Resource start-up disabled since no STONITH resources have been defined
DEBUG: +Event events: error: 2022-11-25T13:55:08.072255+08:00 15sp4-1 pacemaker-schedulerd[4956]:  error: Either configure some or disable STONITH with the stonith-enabled option
DEBUG: +Event events: error: 2022-11-25T13:55:08.072382+08:00 15sp4-1 pacemaker-schedulerd[4956]:  error: NOTE: Clusters with shared data need STONITH to ensure data integrity
DEBUG: {Transition: 15sp4-1:pe-input-0
DEBUG: ~Transition: 15sp4-1:pe-input-0 old(0, pe-input-0.bz2) new(0, pe-input-0.bz2)
DEBUG: }Transition: 15sp4-1:pe-input-0 Complete
DEBUG: {Transition: 15sp4-1:pe-input-1
DEBUG: ~Transition: 15sp4-1:pe-input-1 old(1, pe-input-1.bz2) new(1, pe-input-1.bz2)
DEBUG: }Transition: 15sp4-1:pe-input-1 Complete
DEBUG: +Event quorum: quorum: 2022-11-25T13:58:38.414847+08:00 15sp4-1 pacemaker-controld[4957]:  notice: Updating quorum status to true (call=47)
DEBUG: {Transition: 15sp4-1:pe-input-2
DEBUG: ~Transition: 15sp4-1:pe-input-2 old(2, pe-input-2.bz2) new(2, pe-input-2.bz2)
DEBUG: }Transition: 15sp4-1:pe-input-2 Complete
DEBUG: {Transition: 15sp4-1:pe-input-3
DEBUG: ~Transition: 15sp4-1:pe-input-3 old(3, pe-input-3.bz2) new(3, pe-input-3.bz2)
DEBUG: }Transition: 15sp4-1:pe-input-3 Complete
```

### CODE REVIEW

**Comment by liangxin1300** [comment]
Oh, codeclimate said "Similar blocks"

**Comment by liangxin1300** [comment]
Already changed, dose this make sense?

**Comment by liangxin1300** [comment]
It seems regression test break at history.exp, I guess the input data "history-test.tar.bz2" is old for the new log patterns since regression test environment changed.
I will append a new report data to replace "history-test.tar.bz2"

File: crmsh/utils.py [hunk 1]
```diff
@@ -1245,7 +1252,14 @@ def page_gen(g):
     w, h = get_winsize()
     if not config.core.pager or not can_ask() or options.batch:
         for line in g:
-            sys.stdout.write(term_render(line))
+            try:
+                sys.stdout.write(term_render(line))
+            except UnicodeEncodeError:
```
Comment (liangxin1300) [inline crmsh/utils.py]: When using history subcommand,  due to some special characters like "->" in journal.log, sys.stdout.write will raise UnicodeEncodeError.
If just like #371 did, it also cause "ValueError: I/O operation on closed file."
So according to https://bugs.python.org/issue21363,
I "detach" it, then it works

File: crmsh/utils.py [hunk 2]
```diff
@@ -1245,7 +1252,14 @@ def page_gen(g):
...
+                import io
+                correct_stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')
+                correct_stdout.write(term_render(line))
+                correct_stdout.detach()
+                del correct_stdout
```
Comment (liangxin1300) [inline crmsh/utils.py]: When using history subcommand,  due to some special characters like "->" in journal.log, sys.stdout.write will raise UnicodeEncodeError.
If just like #371 did, it also cause "ValueError: I/O operation on closed file."
So according to https://bugs.python.org/issue21363,
I "detach" it, then it works

File: crmsh/utils.py [hunk 3]
```diff
@@ -1245,7 +1252,14 @@ def page_gen(g):
...
     if not config.core.pager or not can_ask() or options.batch:
         for line in g:
-            sys.stdout.write(term_render(line))
+            try:
+                sys.stdout.write(term_render(line))
+            except UnicodeEncodeError:
+                import io
+                correct_stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')
+                correct_stdout.write(term_render(line))
+                correct_stdout.detach()
```
Comment (liangxin1300) [inline crmsh/utils.py]: When using history subcommand,  due to some special characters like "->" in journal.log, sys.stdout.write will raise UnicodeEncodeError.
If just like #371 did, it also cause "ValueError: I/O operation on closed file."
So according to https://bugs.python.org/issue21363,
I "detach" it, then it works

**Comment by liangxin1300** [comment]
> It seems regression test break at history.exp, I guess the input data "history-test.tar.bz2" is old for the new log patterns since regression test environment changed.
> I will append a new report data to replace "history-test.tar.bz2"

**Comment by liangxin1300** [comment]
> It seems regression test break at history.exp, I guess the input data "history-test.tar.bz2" is old for the new log patterns since regression test environment changed.
> I will append a new report data to replace "history-test.tar.bz2"

Hi, I have updated the report data, just a new "history-test.tar.bz2", is this suitable? Please check.

**Comment by krig** [comment]
Oops, forgot to check this, apologies! Hmm, the history file you uploaded doesn't seem to actually contain any changes, and we should change the history test to actually request transitions that exist. Now I see things like this in the logs:

```
ERROR: 27: no transitions found in the source
```

and then when the test tries to look at transition 49, it doesn't exist. So that is not a very useful test..

File: crmsh/utils.py [hunk 4]
```diff
@@ -1139,6 +1139,13 @@ def is_process(s):
     return False
 
 
+def whether_pacemaker2_daemons():
```
[the remark this answers is not in the document]
Comment (liangxin1300) [reply crmsh/utils.py]: Thanks!

File: crmsh/utils.py [hunk 5]
```diff
@@ -1139,6 +1139,13 @@ def is_process(s):
...
+    for daemon in constants.pacemaker2_daemons:
+        if not os.path.exists(os.path.join("/usr/lib/pacemaker", daemon)):
```
[the remark this answers is not in the document]
Comment (liangxin1300) [reply crmsh/utils.py]: Thank you! Changed

**Comment by liangxin1300** [comment]
> Oops, forgot to check this, apologies! Hmm, the history file you uploaded doesn't seem to actually contain any changes, and we should change the history test to actually request transitions that exist. Now I see things like this in the logs:
> 
> ```
> ERROR: 27: no transitions found in the source
> ```
> 
> and then when the test tries to look at transition 49, it doesn't exist. So that is not a very useful test..

No transitions found also due to patterns, already changed in https://github.com/ClusterLabs/crmsh/pull/428/commits/5d86eabefc57a7203f379d1e3710973035549ad4

**Comment by dmuhamedagic** [comment]
crm history is quite broken, i.e. stuff like history resource doesn't work. This seems to be a good start, but  still doesn't fix it completely. I suppose that this should have a priority a tad higher than low.

**Comment by dmuhamedagic** [comment]
I mean, searching for pacemaker issues is a hell, just as it has always been :-/

**Comment by akoma-2a** [comment]
Hi,
For some reasons I can not get the proper output of the "crm history info" command
First time I ran it gives the right output
...
node1:~ # date; crm history info 
Sat Feb 25 21:58:01 CET 2023
INFO: Retrieving information from cluster nodes, please wait...
Source: live
Created on: Sat Feb 25 21:58:11 CST 2023
By: report -Z -Q -f Sat Feb 25 19:58:01 2023 /var/cache/crm/history-root/live
Period: 2023-02-25 20:58:01 - 2023-02-25 22:58:23
Nodes: node1 node2
Groups: 
Clones: msl_SAPHana_FU2_HDB02 cln_SAPHanaTopology_FU2_HDB02 cl-EmailToAdmin
Resources: stonith-sbd rsc_ip_FU2_HDB02 rsc_SAPHana_FU2_HDB02 rsc_SAPHanaTopology_FU2_HDB02 EmailToAdmin
Transitions: 
node1:~ # 
...

whereas the second time it fails with "RuntimeError: threads can only be started once"
...
node1:~ # date; crm history info 
Sat Feb 25 21:59:47 CET 2023
INFO: Fetching updated logs from cluster nodes. Please wait...
Traceback (most recent call last):
  File "/usr/sbin/crm", line 46, in <module>
    rc = main.run()
  File "/usr/lib/python3.6/site-packages/crmsh/main.py", line 370, in run
    return main_input_loop(context, user_args)
  File "/usr/lib/python3.6/site-packages/crmsh/main.py", line 249, in main_input_loop
    rc = handle_noninteractive_use(context, user_args)
  File "/usr/lib/python3.6/site-packages/crmsh/main.py", line 205, in handle_noninteractive_use
    if context.run(' '.join(l)):
  File "/usr/lib/python3.6/site-packages/crmsh/ui_context.py", line 91, in run
    rv = self.execute_command() is not False
  File "/usr/lib/python3.6/site-packages/crmsh/ui_context.py", line 283, in execute_command
    rv = self.command_info.function(*arglist)
  File "/usr/lib/python3.6/site-packages/crmsh/ui_history.py", line 151, in do_info
    return crm_report().info()
  File "/usr/lib/python3.6/site-packages/crmsh/history.py", line 756, in info
    if not self.prepare_source():
  File "/usr/lib/python3.6/site-packages/crmsh/history.py", line 624, in prepare_source
    self.loc = self.manage_live_report(no_live_update=no_live_update)
  File "/usr/lib/python3.6/site-packages/crmsh/history.py", line 444, in manage_live_report
    rc = self.update_live_report(next_loglines, next_peinputs)
  File "/usr/lib/python3.6/site-packages/crmsh/history.py", line 390, in update_live_report
    rc1 = next_loglines(to_update, self.outdir, self.errdir, end_time)
  File "/usr/lib/python3.6/site-packages/crmsh/crm_pssh.py", line 159, in next_loglines
    statuses = do_pssh(l, opts)
  File "/usr/lib/python3.6/site-packages/crmsh/crm_pssh.py", line 110, in do_pssh
    return manager.run()  # returns a list of exit codes
  File "/usr/lib/python3.6/site-packages/parallax/manager.py", line 102, in run
    writer.start()
  File "/usr/lib64/python3.6/threading.py", line 842, in start
    raise RuntimeError("threads can only be started once")
RuntimeError: threads can only be started once
node1:~ # 
...

Not sure whether it is an expected behavior...

Any help is much appreciated.
Any help is much appreciated.

**Review by zzhou1 — APPROVED** [review]

**Pull request merged**